docs/permission/permission-framework-for-plugin-authors.md

```diff
@@ -0,0 +1,508 @@
+
+    ```javascript
+    import { IdentityClient } from '@backstage/plugin-auth-backend';
+    import { createRouter } from '@internal/plugin-todo-list-backend';
+    import { Router } from 'express';
+    import { PluginEnvironment } from '../types';
+
+    export default async function createPlugin({
+      logger,
+      discovery,
+      permissions,
+    }: PluginEnvironment): Promise<Router> {
+      return await createRouter({
+        logger,
+        identity: new IdentityClient({
+          discovery,
+          issuer: await discovery.getExternalBaseUrl('auth'),
+        }),
+        permissions,
+      });
+    }
+    ```
+
+    Don't worry if you see typescript complaining about the `permissions` object. This is just a typings issue that we will address in the following steps.
+
+    Apply the following changes to `packages/backend/src/index.ts`:
+
+    ```diff
+      import techdocs from './plugins/techdocs';
+    + import todoList from './plugins/todolist';
+      import search from './plugins/search';
+
+      ...
+
+      const searchEnv = useHotMemoize(module, () => createEnv('search'));
+      const appEnv = useHotMemoize(module, () => createEnv('app'));
+    + const todoListEnv = useHotMemoize(module, () => createEnv('todolist'));
+
+      ...
+
+      apiRouter.use('/proxy', await proxy(proxyEnv));
+      apiRouter.use('/search', await search(searchEnv));
+      apiRouter.use('/permission', await permission(permissionEnv));
+    + apiRouter.use('/todolist', await todoList(todoListEnv));
+      // Add backends ABOVE this line; this 404 handler is the catch-all fallback
+      apiRouter.use(notFoundHandler());
+    ```
+
+    Apply the following changes to `packages/app/src/App.ts`:
+
+    ```diff
+    + import { TodoListPage } from '@internal/plugin-todo-list';
+
+    ...
+
+        <Route path="/search" element={<SearchPage />}>
+          {searchPage}
+        </Route>
+        <Route path="/settings" element={<UserSettingsPage />} />
+    +   <Route path="/todo-list" element={<TodoListPage />} />
+      </FlatRoutes>
+    ```
+
+Now if you start your application you should be able to reach the `/todo-list` page:
+
+![Todo List plugin page](../assets/permission/permission-todo-list-page.png)
+
+---
+
+## Integrate the new plugin
+
+If you play with the UI, you will notice that it is possible to perform a few actions:
+
+- create a new todo item (`POST /todos`)
+- view todo items (`GET /todos`)
+- edit an existing todo item (`PUT /todos`)
+
+Let's try to bring authorization on top of each one of them.
+
+## Authorize the create endpoint
+
+The first step we need to do in order to authorize the create endpoint, is to create a new `permission` inside our backend plugin.
+
+Let's create a new `permissions.ts` file under `plugins/todo-list-backend/src/service/permissions.ts` with the following content:
+
+```javascript
+import { Permission } from '@backstage/plugin-permission-common';
```

Comment:
I don't think we should explain what a permission is at this point.
I think plugin authors should already know these concepts. Meaning that, before jumping into the plugin author's section, they should have read the "permission frameworks for adopters". If we explain what a `Permission` is here, we end up having to re-explain something that has been already written in other sections.

What if we mention "make sure you've read the basic concept at this link" at the beginning?

Reply:
Sorry, missed this conversation since it got auto-collapsed by GitHub! 

I think linking to the permission concepts doc will cover some of it, but I still think there's an opportunity to talk about what each property is for - the concepts doc doesn't get into details about the shape of the objects in code, and code is what we're looking at now.

Maybe another option could be to link to both the concepts doc and the [Permission api doc](https://backstage.io/docs/reference/plugin-permission-common.permission)? The api doc is a bit light on details now, but if we backfill some extra doc comments, we should get something with a bit more detail, like [this](https://backstage.io/docs/reference/search-common.indexabledocument).

Reply:
PS - fine of course for the task of adding these links to be a todo! And I'm happy to contribute the permission api doc changes myself.

Reply:
https://github.com/backstage/backstage/pull/9882/ for the permission type api docs.